### PR TITLE
go-backcompat: create back-compat guidance as annotation

### PR DESCRIPTION
**Pull request by bobheadxi**
Generate migration failure warning as an annotation. This should happen rarely enough to warrant taking up the extra space, and I think should primarily happen on PRs?

(aside: I just want to assign a multi line string to a variable but it kept failing 😞 tried a bunch of stack overflows to make this work ugh bash why)

## Test plan

https://buildkite.com/sourcegraph/sourcegraph/builds/133499

image

### CODE REVIEW

File: dev/ci/go-test.sh [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 set -euo pipefail
 
+exit 1
+
```
Comment (bobheadxi) [inline dev/ci/go-test.sh:6]:
```suggestion
```

File: enterprise/dev/ci/internal/ci/operations.go [hunk 2]
```diff
@@ -351,7 +351,7 @@ func addGoTestsBackcompat(minimumUpgradeableVersion string) func(pipeline *bk.Pi
 			pipeline.AddStep(
 				fmt.Sprintf(":go::postgres: Backcompat test (%s)", description),
 				bk.Env("MINIMUM_UPGRADEABLE_VERSION", minimumUpgradeableVersion),
-				bk.Cmd("./dev/ci/go-backcompat/test.sh "+testSuffix),
+				bk.AnnotatedCmd("./dev/ci/go-backcompat/test.sh "+testSuffix, bk.AnnotatedCmdOpts{}),
```
Comment (bobheadxi) [inline enterprise/dev/ci/internal/ci/operations.go:354]: magic sauce

File: dev/ci/go-backcompat/test.sh [hunk 3]
```diff
@@ -136,16 +135,23 @@ fi
...
+${flakefile}
+\`\`\`
+
+Rewrite these schema changes to be backwards compatible. For help,
+see [the migrations guide](docs.sourcegraph.com/dev/background-information/sql/migrations).
+EOF
+  )
+  echo "$annotation"
+  mkdir -p ./annotations/
+  echo "$annotation" >'./annotations/go-backcompat.md'
```
Comment (bobheadxi) [inline dev/ci/go-backcompat/test.sh:155]: the other part of the magic sauce

File: dev/ci/go-backcompat/test.sh [hunk 4]
```diff
@@ -136,16 +135,24 @@ fi
...
-  echo "see docs.sourcegraph.com/dev/background-information/sql/migrations."
+  annotation=$(
+    cat <<EOF
+This commit contains database schema definitions that caused an unexpected
+failure of one or more unit tests at tagged commit \`${latest_minor_release_tag}\`.
+Rewrite these schema changes to be backwards compatible. For help,
+see [the migrations guide](docs.sourcegraph.com/dev/background-information/sql/migrations).
+
+If this backwards incompatibility is intentional or if the test is flaky,
+an exception for this test can be added to the following flakefile:
```
Comment (bobheadxi) [inline dev/ci/go-backcompat/test.sh:146]: I fixed some typos here and also switched it up to recommend writing it in a backwards-compatible way first, _then_ recommend adding to flakefile

**Review by efritz — APPROVED** [review]
Neat!

File: dev/ci/go-backcompat/test.sh (hunk 4 above)
Comment (efritz) [reply dev/ci/go-backcompat/test.sh:146]: Yes, thanks! I noticed that the order wasn't ... great once I discovered an actual flake.

**Pull request merged**